core/metacling/src/TClingCallFunc.cxx

```diff
@@ -1389,384 +1389,85 @@ TClingCallFunc::make_dtor_wrapper(const TClingClassInfo *info)
-               break;
-            case BuiltinType::Float: vh.u.flt = fArgVals[i].simplisticCastAs<float>();
-               break;
-            case BuiltinType::Double: vh.u.dbl = fArgVals[i].simplisticCastAs<double>();
-               break;
-            case BuiltinType::LongDouble: vh.u.ldbl = fArgVals[i].simplisticCastAs<long double>();
-               break;
-            case BuiltinType::NullPtr: vh.u.vp = fArgVals[i].getPtr();
-               break;
-            default:
-               // Thee should be no others.  This is here in case
-               // this changes in the future.
-               ::Error("TClingCallFunc::exec(void*)",
-                       "Unhandled builtin type '%s'",
-                       BT->getTypeClassName());
-                  QT->dump();
-               return;
-            }
-            vh_ary.push_back(vh);
-            vp_ary.push_back(&vh_ary.back());
-         } else if (QT->isReferenceType() || QT->isRecordType()) {
+      for (unsigned i = 0; i < num_args; ++i) {
+         QualType QT;
+         // Check if we provided a this parameter.
+         // FIXME: Currently we do not provide consistently the this pointer as
+         // in C++ we can still call member functions which do not use it. Eg:
+         // struct S {int Print() { return printf("a");} }; auto r1 = ((S*)0)->Print();
+         // This works just fine even though it might be UB...
+         if (i == 0 && isa<CXXMethodDecl>(FD) && num_args - FD->getNumParams() == 1)
+           QT = cast<CXXMethodDecl>(FD)->getThisType();
+         else
+           QT = FD->getParamDecl(i)->getType();
+         QT = QT.getCanonicalType();
+         if (QT->isReferenceType() || QT->isRecordType()) {
             // the argument is already a pointer value (points to the same thing
             // as the reference or pointing to object passed by value.
-            vp_ary.push_back((void *) fArgVals[i].simplisticCastAs<unsigned long long>());
-         } else if (QT->isPointerType() || QT->isArrayType() || QT->isMemberPointerType()) {
-            ValHolder vh;
-            vh.u.vp = (void *) fArgVals[i].simplisticCastAs<unsigned long long>();
-            vh_ary.push_back(vh);
-            vp_ary.push_back(&vh_ary.back());
-         } else if (QT->isEnumeralType()) {
-            // FIXME: Handle isScopedEnumeralType ()
-            // FIXME:We may need to worry about the underlying type of the enum
-            ValHolder vh;
-            vh.u.i = fArgVals[i].simplisticCastAs<int>();
-            vh_ary.push_back(vh);
-            vp_ary.push_back(&vh_ary.back());
-         } else {
-            ::Error("TClingCallFunc::exec(void*)",
-                    "Invalid type '%s'", QT->getTypeClassName());
-            QT->dump();
-            return;
-         }
+            vp_ary.push_back(fArgVals[i].getPtr());
+         } else
+            vp_ary.push_back(&fArgVals[i].getPtr());
```

Comment:
Once the const-overload gets called here (e.g. because in a future version we mark this function `const`) this will return the address to a temporary. We can hope for the compiler to complain - maybe make this more explicitly the non-const call, or add a `static_assert`?

Reply:
Now we have a `getPtrAddress` which makes the operation more explicit. I believe that addresses your concern but please take a look again.